spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -747,58 +788,73 @@ struct ProductionVisitor : public production::Visitor {
 
         if ( length ) {
             // Limit input to the specified length.
-            auto limited = builder()->addTmp("limited_", builder()->memberCall(state().cur, "limit", {*length}));
+            auto limited = builder()->addTmp("limited_", builder()->memberCall(state().cur, "limit", {length}));
 
             // Establish limited view, remembering position to continue at.
             auto pstate = state();
             pstate.cur = limited;
-            pstate.ncur = builder()->addTmp("ncur", builder()->memberCall(state().cur, "advance", {*length}));
+            pstate.ncur = builder()->addTmp("ncur", builder()->memberCall(state().cur, "advance", {length}));
             pushState(std::move(pstate));
         }
         else {
             auto pstate = state();
             pstate.ncur = {};
             pushState(std::move(pstate));
         }
+    }
 
-        if ( pb->options().getAuxOption<bool>("spicy.track_offsets", false) ) {
-            auto __offsets = builder()->member(state().self, "__offsets");
-            auto cur_offset = builder()->memberCall(state().cur, "offset");
+    // Tears down parsing infrastructure for non-optimized fields.
+    //
+    // Returns two expressions:
+    // 1. `cur` from the state that was current when calling the method.
+    // 2. `ncur_max_size` if we're parsing with `&max-size`, which in that case
+    // is the position in the limited view we ended parsing to; this will be
+    // used to compute how much data we consumed from the original view.
+    std::pair<Expression*, Expression*> finishStandardFieldParsing(type::unit::item::Field* field) {
+        std::pair<Expression*, Expression*> result = {state().ncur, {}};
```

Comment:
> There are a lot of accesses to state() in the body which IMO makes this code harder to follow.

I think it's the opposite: it makes it easier because it means "use the current state", i.e., one doesn't need to worry if there was push/pop state somehwere in between.